### ISSUE TITLE
Checkpointing and loading training states

### ISSUE THREAD

**Issue by Dzz99** [neuraloperator/neuraloperator#480]
I encountered an issue while running the "Checkpointing and loading training states" example from the user manual. After using resume_from_dir, the model is not training properly.

Here is the code I used:

device = "cuda" if torch.cuda.is_available() else "cpu"
# device ="cpu"
train_loader, test_loaders, data_processor = load_darcy_flow_small(
        n_train=1000, batch_size=32,
        test_resolutions=[16, 32], n_tests=[100, 50],
        test_batch_sizes=[32, 32],
)

model = FNO(n_modes=(4,4),
             in_channels=1,
             out_channels=1,
             hidden_channels=4,
             projection_channel_ratio=1,
             factorization='tucker',
             rank=0.42)

model = model.to(device)
optimizer = AdamW(model.parameters(),
                  lr=8e-3,
                  weight_decay=1e-4)
scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, eta_min=8e-3, T_max=30)
l2loss = LpLoss(d=2, p=2)
h1loss = H1Loss(d=2)

train_loss = h1loss
eval_losses = {'h1': h1loss, 'l2': l2loss}

trainer = Trainer(model=model, n_epochs=10,
                  device=device,
                  data_processor=data_processor,
                  wandb_log=False,
                  eval_interval=1,
                  use_distributed=False,
                  verbose=True)
trainer.train(train_loader=train_loader,
              test_loaders={},
              optimizer=optimizer,
              scheduler=scheduler,
              regularizer=False,
              training_loss=train_loss,
              save_every=7,
              save_dir="./checkpoints")

# .. resume_from_dir:
trainer = Trainer(model=model, n_epochs=10,
                  device=device,
                  data_processor=data_processor,
                  wandb_log=False,
                  eval_interval=1,
                  use_distributed=False,
                  verbose=True)
trainer.train(train_loader=train_loader,
              test_loaders={},
              optimizer=optimizer,
              scheduler=scheduler,
              regularizer=False,
              training_loss=train_loss,
              resume_from_dir="./checkpoints")
The code for saving the model state works fine, but after calling resume_from_dir, the model parameters do not update during training. Based on my testing, it seems that the issue lies in the train_one_epoch function in trainer.py, where the command self.optimizer.step() does not seem to have any effect.

**Comment by dhpitt** [comment]
Hi @Dzz99 , thanks for bringing this to our attention. How recently have you pulled the newest code from our repo? We made some changes to the Trainer last month that should affect this.

**Comment by ybangaru** [comment]
I may have stumbed on the same issue, I pulled the code today as well before testing this

`
2024-12-17 16:31:21.055 | INFO     | neuralop.training.trainer:on_epoch_start:392 - Starting epoch 7
[7] time=76.03, avg_loss=0.9989, train_err=1.9785
Eval: -1_h1=0.9991, -1_l2=1.0324
[Rank 0]: saved training state to /data/mlruns/240161569526831615/9697a49247e34742b9777afbd49a6941/artifacts/model_checkpoints
2024-12-17 16:32:58.872 | INFO     | neuralop.training.trainer:on_epoch_start:392 - Starting epoch 8
[8] time=74.91, avg_loss=0.9989, train_err=1.9785
Eval: -1_h1=0.9991, -1_l2=1.0324
[Rank 0]: saved training state to /data/mlruns/240161569526831615/9697a49247e34742b9777afbd49a6941/artifacts/model_checkpoints
2024-12-17 16:34:35.389 | INFO     | neuralop.training.trainer:on_epoch_start:392 - Starting epoch 9
[9] time=75.06, avg_loss=0.9989, train_err=1.9785
Eval: -1_h1=0.9991, -1_l2=1.0324
[Rank 0]: saved training state to /data/mlruns/240161569526831615/9697a49247e34742b9777afbd49a6941/artifacts/model_checkpoints
2024-12-17 16:36:11.928 | INFO     | neuralop.training.trainer:on_epoch_start:392 - Starting epoch 10
[10] time=75.07, avg_loss=0.9989, train_err=1.9785
Eval: -1_h1=0.9991, -1_l2=1.0324
[Rank 0]: saved training state to /data/mlruns/240161569526831615/9697a49247e34742b9777afbd49a6941/artifacts/model_checkpoints
2024-12-17 16:37:48.506 | INFO     | neuralop.training.trainer:on_epoch_start:392 - Starting epoch 11
[11] time=75.09, avg_loss=0.9989, train_err=1.9785
Eval: -1_h1=0.9991, -1_l2=1.0324
[Rank 0]: saved training state to /data/mlruns/240161569526831615/9697a49247e34742b9777afbd49a6941/artifacts/model_checkpoints
2024-12-17 16:39:25.211 | INFO     | neuralop.training.trainer:on_epoch_start:392 - Starting epoch 12
[12] time=75.33, avg_loss=0.9989, train_err=1.9785
`

**Comment by dhpitt** [comment]
Hi @Dzz99 , @ybangaru , confirming that I can reproduce a similar bug in my local env. I'm looking into this and hope to push a fix shortly.

**Comment by ybangaru** [comment]
Hey @dhpitt , just a heads up, I believe that the error is being propogated from `from_checkpoint` method of the `BaseModel` class, because loading the model state directly using `load_state_dict` does update the training state after each epoch

in training_state.py, with the following change basically
```
    else:
        # model = model.from_checkpoint(save_dir.absolute().as_posix(), save_name, map_location=map_location)
        save_pth = save_dir / f"{save_name}_state_dict.pt"
        model.load_state_dict(torch.load(save_pth.absolute().as_posix(), map_location=map_location))
```

also, while you are at it, the type hints of `load_training_state` and `save_training_state` could use minor changes as optimizer, scheduler and regulalizer may not be of type `nn.Module` , perhaps wont make much difference but just thought you should know :) Thank you very much!

**Comment by dhpitt** [comment]
Thanks @ybangaru , the issue ended up being slightly more complicated, PR is here: #489

**Comment by JeanKossaifi** [comment]
Thanks @ybangaru @dhpitt -  indeed the `from_checkpoint` should be a class method and returns a new instance, we should be loading the state-dict in the existing model instead. Closing as fixed in https://github.com/neuraloperator/neuraloperator/pull/491, we'll make the interface a little smoother going forward. Feel free to reopen if you still have issues.
